### ISSUE TITLE
feat(tune): end-to-end LoRA fine-tuning pipeline

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#88]
## Context

PR #82 added the low-level `train_lora(adapter, samples, config)` API that performs gradient steps given pre-computed `TrainSample { layer_idx, module, input, target_delta }` pairs. It is **not** end-to-end — callers must supply the activations and deltas themselves.

What does NOT exist:
- Forward-pass activation capture at LoRA injection points
- Cross-entropy loss on next-token prediction
- Backprop engine from output logits through the frozen transformer to LoRA A/B matrices
- Higher-level API that takes `(prompt, target_response)` pairs
- Metal LoRA backward pass

Result: there's no way to take a model + text data + train a useful LoRA. The unit tests pass but the feature is unusable from a consumer perspective.

PR #82 was closed because shipping the inner step without the outer loop is misleading.

## Goal

Build the end-to-end pipeline so a downstream consumer can:

```rust
let mut model = Qwen35Model::load(&path)?;
let mut adapter = LoraAdapter::new(LoraConfig { rank: 8, alpha: 16.0, target_modules: vec!["q_proj", "v_proj"] });
let samples: Vec<(String, String)> = load_sft_dataset()?;
train_lora_from_text(&mut model, &mut adapter, &samples, &config)?;
adapter.save_safetensors("adapter.safetensors")?;
```

### 1. Activation capture mode
Extend the forward pass to optionally record activations at LoRA injection points (q_proj, k_proj, v_proj, gate_proj, up_proj, down_proj inputs). New parameter on `BertModel::forward` / `Qwen35Model::forward` or a sibling method.

### 2. Cross-entropy loss + label shifting
Standard next-token prediction loss. Input: model logits + target token IDs. Output: scalar loss + per-token gradients on logits.

### 3. Backprop through frozen transformer to LoRA
The non-LoRA weights are frozen, so we only need gradients flowing backward to the LoRA A/B matrices. The math:
- Loss → ∂L/∂logits (standard CE backward)
- ∂L/∂logits → ∂L/∂hidden (last layer linear backward, frozen weights → only need to compute, not store, weight grads)
- Continue backward through each layer
- At each LoRA injection point: ∂L/∂(B·A·x) splits into ∂L/∂A and ∂L/∂B

This is essentially a small autograd engine scoped to the LoRA-only graph. Reuses the existing `compute_lora_gradients` (from PR #82's optimizer.rs) for the final step.

### 4. Higher-level API: `train_lora_from_text`
Orchestrates: tokenize prompts → forward with activation capture → compute deltas via backprop → call `train_lora` inner step → repeat.

### 5. Metal LoRA backward pass (optional Phase 2)
For training on Metal GPU. CPU-only path is the v1 deliverable.

## Acceptance

- Integration test: load Qwen3.5-0.8B Q4 QuaRot, train a small adapter on 100 (prompt, response) pairs, verify loss decreases over epochs
- Round-trip: save adapter → load via Metal → generate with vs without adapter → outputs diverge measurably
- See also #89 for quality comparison against PyTorch

## Estimated scope

~1-2 weeks of focused work. Probably ADR-058 (after current ADR-058 is renumbered) explaining the design.

## References

- PR #82 (closed) — introduced low-level `train_lora`
- `optimizer.rs` — `AdamState`, `compute_lora_gradients` (will be in PR for standalone landing)
- ADR-008 — LoRA injection trait
- ADR-057 — LoRA full-lifecycle consumer API (defines the consumer shape)
- ADR-033 — JIT adaptation (currently placeholder, per §Consequences/Negative)
- `crates/tune/src/train/jit.rs:251` — `// Simulate training step (placeholder)`

**Comment by ohdearquant** [comment]
**Progress (2026-06-23):** exact-gradient LoRA backward + multi-layer CPU training pipeline shipped (PR #191); adapter alpha now loaded from the safetensors header (PR #261). Driven via `train_grad_full`. Remaining: library API (`train_lora_from_text`), Metal backward pass, HF dataset loading wrapper.

**Comment by ohdearquant** [comment]
The end-to-end pipeline this epic asked for now exists: train_micro_lora(model, pairs, config) -> LoraAdapter (crates/tune/src/lora/train.rs:234) takes tokenized (prompt, completion) TrainingPair sequences, computes cross-entropy next-token loss with exact CPU gradients through the frozen transformer suffix (activation tape covers the deployed suffix from first_layer), trains q_proj/v_proj A/B, and LoraAdapter::save_safetensors (lora/mod.rs:142) writes the adapter. Design and limits: docs/lora-core.md. The remaining gaps are tracked in narrower issues: #737 (large-corpus residency / scalar backward performance, incl. the absent Metal backward), #615 (shared adapter descriptor between tune and inference), #884 (GDN-vector training coverage).
